### DISCUSSION TITLE
Add count subquery for joined entity of parent entity

### DISCUSSION THREAD

**Discussion by bryan-vh** [mikro-orm/mikro-orm#2452] [Q&A] [answered]
Hello, I am trying to add a count subquery for a joined entity (User) of its parent entity (Party) when querying as seen in the code below:

```
const qb = this.em.createQueryBuilder(Party, 'p')

const friendsCountQb = this.em
      .createQueryBuilder(Relationship, 'r')
      .count('*')
      .as('u__friends_count')

qb.select(['*', friendsCountQb])
      .joinAndSelect('p.user', 'u')
      .limit(10)

const results = qb.getResultList()

return results
```

I'm not sure why, but when I look at the results from the query it seems to drop the friendsCount property. It should be noted I have this property on the User entity as { persist: false } so not sure why it wouldn't be showing up. Any help would be great!

**Comment by B4nan** [comment]
It is not enough to just select something with alias that would fit the naming - the QB instance does not know what you are trying to do just because there is such field. 

Will need to try to see if you could hack it somehow. Right now I'd say this is rather not supported, you might be albe to do the same with `@Formula` most probably.

> when I look at the results from the query it seems to drop the friendsCount property. 

Not dropping, it simply ignores it, which is the default behaviour with `qb.getResult()`, you should see it with `qb.execute()`. But that wouldn't give you entity instances...

~Btw your query does not make much sense, you can't use `limit` with to-many joins, that won't work as you expect, as the limit is applied to the whole result set which has more lines than number of selected root entities. You can use `QueryFlag.PAGINATE` for that which will wrap things in subqueries to get around such issues.~ edit: Actually not, I see you are joining `p.user`, I though it was `p.users`

**Reply by bryan-vh** [reply]
Could I use qb.execute() and then map it using the map function to actually convert them into entities? I tried running this same query with raw SQL and it got all the properties that I seemingly wanted but I didn't know a good way to convert them into the appropriate entities after getting all the columns I wanted.

**Reply by B4nan** [reply]
You could try, this is basically what the `qb.getResult()` does:

```ts
res = em.getDriver().mergeJoinedResult(res, this.metadata.find(entityName)!);
return res.map(r => em.map(entityName, r));
```

But I would be surprised if that will work (you would execute the same methods as you already do, just doing some stuff explicitly).

What could work is if you make sure the input for `em.map()` will have the new property named as the virtual property, not sure what the `qb.execute()` gives you.

**Reply by bryan-vh** [reply]
So as of now, there's no way to do this with query builder or knex? Or I'd just have to try and hack it with the formula decorator?

**Reply by B4nan** [reply]
As said above, you can map it yourself before calling `em.map()`. `qb.execute()` will give you the results, `mergeJoinedResult` will convert them to the object graph correctly, you just need to handle remapping of the additional column to the right name.

**Comment by bryan-vh** [comment]
```
const raw = await qb.execute()
const res = this.em.getDriver().mergeJoinedResult(raw, this.em.getMetadata().find('Party')!)

const mapped = res.map((p) => {
      const party = this.em.map(Party, p)
      party.user.friendsCount = p.friends_count
      return party
})

return mapped
```

@B4nan thank you for the help, this is what worked for me! I am curious if there is a better way to do mergeJoinedResult without the non-null assertive operator but otherwise looks good :)

**Reply by B4nan** [reply]
you can do `this.em.getMetadata().get('Party')` which throws if such entity is not discovered. internally `find` is used as a microoptimization

**Reply by bryan-vh** [reply]
@B4nan when I try to run the following query using knex and mapping it similar to how the query builder does it, I can't seem to get the appropriate entity in the response:

```
const query = knex
      .select([
        'p.*',
        'u.id as u__id',
        'u.username as u__username',
        'u.name as u__name',
        'u.birthday as u__birthday',
        'u.profile as u__profile',
        'u.photos as u__photos'
      ])
      .from('party as p')
      .innerJoin('user as u', 'u.id', 'p.user_id')

const raw = await this.em.execute(query)
console.log(raw)
const res = this.em.getDriver().mergeJoinedResult(raw, this.em.getMetadata().find('Party')!)
console.log(res)
```

This is what I get in the JSON response:

```
[
  {
    id: 'BE3AD253-F5A7-4A7D-90A0-A6CBB3E6A8E8',
    name: 'turtle sesh',
    image: 'https://firebasestorage.googleapis.com:443/v0/b/fiesta-97e81.appspot.com/o/212252FF-847D-4D2A-8C6D-E4D0E19E54A9?alt=media&token=9eddfb11-7db2-4e41-ae45-da4007ffaa76',
    start: 2021-11-07T22:30:38.000Z,
    end: 2021-11-08T01:30:38.000Z,
    latitude: 48,
    longitude: -122,
    geog: '0101000020E6100000A05D746847825EC04A0F9E6406CD4740',
    privacy: 0,
    address: 'Sammamish, WA, United States',
    description: null,
    user_id: '4LvcX3AHjRh1FfwD4ccEX7VfnXu2',
    u__id: '4LvcX3AHjRh1FfwD4ccEX7VfnXu2',
    u__username: 'bryan',
    u__name: 'Bryan Van Horn',
    u__birthday: 1999-09-01T07:00:00.000Z,
    u__profile: 'https://firebasestorage.googleapis.com:443/v0/b/fiesta-97e81.appspot.com/o/B7E89BE9-8703-458A-8E36-E1ED39B450A8?alt=media&token=2d5485d7-62d0-425d-94f0-e618cc926bce',
    u__photos: []
  },
  {
    id: '1C1B2671-F603-4280-AE33-609F9FF2F6E2',
    name: 'Japanese Garden Party',
    image: 'https://firebasestorage.googleapis.com:443/v0/b/fiesta-97e81.appspot.com/o/030634E4-7F96-4206-9F0B-C3E8CDF4DAE9?alt=media&token=2600131d-0182-48f9-a640-0d9de02bbe5e',
    start: 2021-11-13T05:30:46.000Z,
    end: 2021-11-13T07:30:46.000Z,
    latitude: 34,
    longitude: -118,
    geog: '0101000020E61000007D8C5EB2169D5DC0259012BBB6054140',
    privacy: 0,
    address: '11613 Santa Monica Blvd, Los Angeles, CA  90025, United States',
    description: null,
    user_id: '4LvcX3AHjRh1FfwD4ccEX7VfnXu2',
    u__id: '4LvcX3AHjRh1FfwD4ccEX7VfnXu2',
    u__username: 'bryan',
    u__name: 'Bryan Van Horn',
    u__birthday: 1999-09-01T07:00:00.000Z,
    u__profile: 'https://firebasestorage.googleapis.com:443/v0/b/fiesta-97e81.appspot.com/o/B7E89BE9-8703-458A-8E36-E1ED39B450A8?alt=media&token=2d5485d7-62d0-425d-94f0-e618cc926bce',
    u__photos: []
  }
]
[
  {
    id: 'BE3AD253-F5A7-4A7D-90A0-A6CBB3E6A8E8',
    name: 'turtle sesh',
    image: 'https://firebasestorage.googleapis.com:443/v0/b/fiesta-97e81.appspot.com/o/212252FF-847D-4D2A-8C6D-E4D0E19E54A9?alt=media&token=9eddfb11-7db2-4e41-ae45-da4007ffaa76',
    start: 2021-11-07T22:30:38.000Z,
    end: 2021-11-08T01:30:38.000Z,
    latitude: 48,
    longitude: -122,
    geog: '0101000020E6100000A05D746847825EC04A0F9E6406CD4740',
    privacy: 0,
    address: 'Sammamish, WA, United States',
    description: null,
    user_id: '4LvcX3AHjRh1FfwD4ccEX7VfnXu2',
    u__id: '4LvcX3AHjRh1FfwD4ccEX7VfnXu2',
    u__username: 'bryan',
    u__name: 'Bryan Van Horn',
    u__birthday: 1999-09-01T07:00:00.000Z,
    u__profile: 'https://firebasestorage.googleapis.com:443/v0/b/fiesta-97e81.appspot.com/o/B7E89BE9-8703-458A-8E36-E1ED39B450A8?alt=media&token=2d5485d7-62d0-425d-94f0-e618cc926bce',
    u__photos: []
  },
  {
    id: '1C1B2671-F603-4280-AE33-609F9FF2F6E2',
    name: 'Japanese Garden Party',
    image: 'https://firebasestorage.googleapis.com:443/v0/b/fiesta-97e81.appspot.com/o/030634E4-7F96-4206-9F0B-C3E8CDF4DAE9?alt=media&token=2600131d-0182-48f9-a640-0d9de02bbe5e',
    start: 2021-11-13T05:30:46.000Z,
    end: 2021-11-13T07:30:46.000Z,
    latitude: 34,
    longitude: -118,
    geog: '0101000020E61000007D8C5EB2169D5DC0259012BBB6054140',
    privacy: 0,
    address: '11613 Santa Monica Blvd, Los Angeles, CA  90025, United States',
    description: null,
    user_id: '4LvcX3AHjRh1FfwD4ccEX7VfnXu2',
    u__id: '4LvcX3AHjRh1FfwD4ccEX7VfnXu2',
    u__username: 'bryan',
    u__name: 'Bryan Van Horn',
    u__birthday: 1999-09-01T07:00:00.000Z,
    u__profile: 'https://firebasestorage.googleapis.com:443/v0/b/fiesta-97e81.appspot.com/o/B7E89BE9-8703-458A-8E36-E1ED39B450A8?alt=media&token=2d5485d7-62d0-425d-94f0-e618cc926bce',
    u__photos: []
  }
]
```

As you can see if I run this without query builder and the mapResults boolean that comes with its execute function I can't seem to get the properties to nest properly. Any help with this would be really helpful! If there's any way where I wouldn't have to assign the properties manually that would be great as that's what I'm looking to avoid (if possible). **It should be noted I have to use knex for future queries so I cannot revert to query builder.** Is it possible to use the driver's `mapResult` function?

**Reply by B4nan** [reply]
You can try to play with `mapResult()` explicitly, you still need the metadata that tells the ORM what relations are joined to what alias. It would be best if you try to debug how the QB works and mimic that, I can't really give you any example from the top of my head. Those are all internal things, and given you want to work with raw knex, it will be all even more tedious, curious why you need that?

**Reply by bryan-vh** [reply]
@B4nan, I've been able to write the following query with knexjs but not with query builder so curious how I'd do it:

```
const query = knex
      .select([
        'p.*',
        'u.id as u__id',
        'u.username as u__username',
        'u.name as u__name',
        'u.birthday as u__birthday',
        'u.profile as u__profile',
        'u.photos as u__photos'
      ])
      .from('party as p')
      .innerJoin(
        knex
          .select('first_id')
          .from('relationship')
          .where('second_id', id)
          .andWhere('type', RelationshipType.Friends)
          .union(
            knex
              .select('second_id')
              .from('relationship')
              .where('first_id', id)
              .andWhere('type', RelationshipType.Friends)
          )
          .as('r'),
        'r.first_id',
        'p.user_id'
      )
      .innerJoin('user as u', 'u.id', 'p.user_id')
      .whereIn('p.privacy', [Privacy.Public, Privacy.Friends])
```

In the query I'm trying to do an inner join based on a select statement. I'm pretty new to SQL so if there's a better way to improve upon this query or way to do it with query builder that would be amazing. Thanks for all the help, but that's why I'm using knex over the qb at the moment :)

**Reply by B4nan** [reply]
Now that is a very complex query. You could probably left join with multiple on conditions, I don't see a need for this complexity. Also that inner union does not make any sense to me.

But please don't expect me to deal with your app problems, it is quite time consuming for me to understand your needs based on such complex queries.

**Reply by bryan-vh** [reply]
@B4nan totally agree, that's why I didn't want to fully delve into why I needed to use knex, just wanted to understand the difference between using one vs the other when it comes to getting proper results. Thank you anyhow!

**Comment by bryan-vh** [comment, marked as answer]
Solution using query builder after realizing I overthought my SQL statement that worked for me thanks to @B4nan's help!

```
const qb = this.em.createQueryBuilder(Party, 'p')

const likesCountQb = this.em
      .createQueryBuilder(Like, 'l')
      .count('*')
      .where({ party: knex.ref('p.id') })
      .as('likes_count')

const reservationsCountQb = this.em
      .createQueryBuilder(Reservation, 'r')
      .count('*')
      .where({ party: knex.ref('p.id') })
      .as('reservations_count')

const friendsCountQb = this.em
      .createQueryBuilder(Relationship, 'r')
      .count('*')
      .where('r.first_id = u.id and r.type = 2')
      .orWhere('r.second_id = u.id and r.type = 2')
      .as('friends_count')

qb.select(['*', likesCountQb, reservationsCountQb, friendsCountQb])
      .joinAndSelect('p.user', 'u')
      .join('u.firstRelationships', 'r1')
      .join('u.secondRelationships', 'r2')
      .where('r1.second_id = ? and r1.type = 2', [id])
      .orWhere('r2.first_id = ? and r2.type = 2', [id])
      .andWhere({ 'p.privacy': [Privacy.Public, Privacy.Friends] })
      .limit(10)

const raw = await qb.execute()
const results = raw.map((p) => {
      const party = this.em.map(Party, p)
      party.user.friendsCount = p.friends_count
      return party
})

return results
```
